### ISSUE TITLE
Upgrade to Scala 3

### ISSUE THREAD

**Issue by adpi2** [scalacenter/scaladex#1104]
### Motivation

Compile with Scala 3

### Identified Obstacles

- We will probably need to make some major bump of some libraries and fix the breaking changes.
- Some libraries have not yet been published for Scala 3 but we can try to use `CrossVersion.for3Use2_13`.
- Some libraries may not be compatible with Scala 3 even with `CrossVersion.for3Use2_13`.

### Implementation Guideline

- [ ] Starting from `core` module
  - [ ] Upgrade the libraries to some Scala 3 compatible versions and fix all breaking changes.
  - [ ] Use `CrossVersion.for3Use2_13` for libraries that are not yet compatible with Scala 3.
  - [ ] Change Scala version of the module to Scala 3
  - [ ] If `CrossVersion.for3Use2_13` does not work it can be because of conflicting transitive dependencies, or it could be that we need to use Macros. Can we resolve those conflicting dependencies manually? Can we stop using the macros?
  - [ ] Run the tests
- [ ] Start again with another module like `infra` or `client` until every module compiles to Scala 3.

### Expectations

- All tests should pass using Scala 3

**Comment by mfirry** [comment]
I deleted the comment, but thanks @adpi2 for your reply. I'll think about this.

**Comment by adpi2** [comment]
Actually we should have a license for Akka soon. I'll keep you updated about it.

**Comment by mfirry** [comment]
Thanks @adpi2. I did give it a try to "just" work on _core_ (and the easy way for me was to sort of split it up on its own repo).
With the precious help from @TonioGela, we made it compile. Please have a look in your own good time https://github.com/mfirry/scaladex-core 

As far as I can see all other modules need Akka.

**Comment by adpi2** [comment]
> Thanks @adpi2. I did give it a try to "just" work on core (and the easy way for me was to sort of split it up on its own repo).
With the precious help from @TonioGela, we made it compile

Yay, that looks beautiful!

So I can confirm that we have a license for Akka but it is a temporary one. I would rather try to migrate to Pekko than to renew our Akka license every year.

**Comment by mfirry** [comment]
Ok. How do you suggest one can go ahead? What order would you suggest projects should be worked? I see these: _**infra, webclient, server, core, data**_? Am I missing anything?

**Comment by adpi2** [comment]
> Ok. How do you suggest one can go ahead? What order would you suggest projects should be worked? I see these: infra, webclient, server, core, data? Am I missing anything?

Here is the migration to Pekko: https://github.com/scalacenter/scaladex/pull/1338

Before starting the migration I would check that every library is Scala 3 compatible, and update them if needed. For the one that are not Scala 3 compatible I would configure them with `CrossVersion.for3Use2_13`.

Then I would suggest this order of projects: **core, webclient, infra, data, server**.

**Comment by adpi2** [comment]
> Here is the migration to Pekko: https://github.com/scalacenter/scaladex/pull/1338

It's merged

**Comment by mfirry** [comment]
Sorry... got lost in so many things. Has anyone moved in any direction here? If not, I'll try and make a proper step-by-step upgrade during the summer. Let me know @adpi2 and thanks.

**Comment by mfirry** [comment]
Ah cool. Anything public I can see and follow?

**Comment by mfirry** [comment]
Hi @kannupriyakalra ... any update?
